jxl/src/render/stages/noise.rs

```diff
@@ -0,0 +1,357 @@
+        position: (usize, usize),
+        xsize: usize,
+        row: &mut [&mut [f32]],
+    ) {
+        let noise_model = StrengthEvalLut { noise: &self.noise };
+        let norm_const = 0.22;
+        // TODO(zond): use self.color_correlation.
+        let ytox = 0.0;
+        // TODO(zond): use self.color_correlation.
+        let ytob = 1.0;
+        for x in 0..xsize {
+            let row_rnd_r = row[self.first_channel][x];
+            let row_rnd_g = row[self.first_channel + 1][x];
+            let row_rnd_c = row[self.first_channel + 2][x];
+            let vx = row[0][x];
+            let vy = row[1][x];
+            let in_g = vy - vx;
+            let in_r = vy + vx;
+            let noise_strength_g = noise_strength(&noise_model, in_g * 0.5);
+            let noise_strength_r = noise_strength(&noise_model, in_r * 0.5);
+            let addit_rnd_noise_red = row_rnd_r * norm_const;
+            let addit_rnd_noise_green = row_rnd_g * norm_const;
+            let addit_rnd_noise_correlated = row_rnd_c * norm_const;
+            println!(
+                "{}, {}, {}, {}, {}, {}, {}",
+                addit_rnd_noise_red,
+                addit_rnd_noise_green,
+                addit_rnd_noise_correlated,
+                noise_strength_g,
+                noise_strength_r,
+                ytox,
+                ytob
+            );
+            let k_rg_corr = 0.9921875f32;
+            let k_rgn_corr = 0.0078125f32;
+            let red_noise = noise_strength_r
+                * (k_rgn_corr * addit_rnd_noise_red + k_rg_corr * addit_rnd_noise_correlated);
+            let green_noise = noise_strength_g
+                * (k_rgn_corr * addit_rnd_noise_green + k_rg_corr * addit_rnd_noise_correlated);
+            let rg_noise = red_noise + green_noise;
+            row[0][x] += ytox * rg_noise + red_noise - green_noise;
+            row[1][x] += rg_noise;
+            row[2][x] += ytob * rg_noise;
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::{
+        error::Result,
+        features::noise::Noise,
+        image::Image,
+        render::{
+            stages::noise::{AddNoiseStage, ConvolveNoiseStage},
+            test::make_and_run_simple_pipeline,
+            RenderPipelineStage,
+        },
+        util::test::assert_almost_eq,
+    };
+    use test_log::test;
+
+    #[test]
+    fn convolve_noise_uses_channels() -> Result<()> {
+        let stage = ConvolveNoiseStage::new(1);
+        assert!(
+            !stage.uses_channel(0),
+            "shouldn't use channel < first_channel"
+        );
+        for c in 1..=3 {
+            assert!(
+                stage.uses_channel(c),
+                "should use first_channel <= channel < first_channel + 3"
+            );
+        }
+        assert!(
+            !stage.uses_channel(4),
+            "shouldn't use channel > first_channel + 3"
+        );
+        Ok(())
+    }
+
+    #[test]
+    fn convolve_noise_process_row_chunk() -> Result<()> {
```

Comment:
I would rather test properties such as "doing the noise convolution on a constant image produces an all-0 image" rather than specific values.

Reply:
Also testing the impulse response on a 5x5 all-0 image except for the center pixel which is 1 seems a reasonable test.

Reply:
Added TODO(mo271) for this as per Discord discussion.